### PR TITLE
Add Support for Knowledgeable Passage Retriever (KPR)

### PR DESCRIPTION
**Pull request by ikuyamada**
Hello,  

This pull request adds support for our new **Knowledgeable Passage Retriever (KPR)** model, which will be presented at the upcoming EMNLP conference.  
KPR improves embedding models on queries involving less-frequent entities by injecting external entity knowledge through an attention mechanism.

Paper: https://arxiv.org/abs/2507.03922  
Code: https://github.com/knowledgeable-embedding/knowledgeable-embedding  

This change is necessary because our model requires additional inputs: `entity_ids`, `entity_position_ids`, and `entity_embeds`, in addition to the standard features such as `input_ids` and `attention_mask`.  

Thanks in advance for reviewing this PR!

### CODE REVIEW

**Comment by tomaarsen** [comment]
Hello!

Very nice work, and solid work getting this so nicely integrated with Transformers. I'd be glad to help get this integrated in Sentence Transformers as well to simplify the usage even further.

However, I think we can avoid going for a architecture-specific implementation with specifying `entity_ids`, etc. Instead, we can perhaps look at the signature of the underlying model and pass any arguments that match whatever the underlying model requires?

Feel free to have a look at my commit that makes this change.

It won't work out of the box yet, because your [modeling code](https://github.com/knowledgeable-embedding/knowledgeable-embedding/blob/bd4467195fea83866c6d8bcbf7049e08d8c70a38/src/kembed/modeling.py#L222-L251) (reasonably) uses `*args, **kwargs` in its signature instead. If you're open to it, you can update this with e.g.:
```python
class KPRModelForBert(BertPreTrainedModel, KPRMixin):
    config_class = KPRConfigForBert

    def __init__(self, config: KPRConfigForBert):
        BertPreTrainedModel.__init__(self, config)

        self.bert = BertModel(config)
        if self.config.entity_fusion_method != "none":
            self.entity_fusion_layer = EntityFusionLayer(config)

        self.post_init()

    def forward(
        self,
        input_ids: Optional[torch.Tensor] = None,
        attention_mask: Optional[torch.Tensor] = None,
        token_type_ids: Optional[torch.Tensor] = None,
        position_ids: Optional[torch.Tensor] = None,
        head_mask: Optional[torch.Tensor] = None,
        inputs_embeds: Optional[torch.Tensor] = None,
        output_attentions: Optional[bool] = None,
        output_hidden_states: Optional[bool] = None,
        return_dict: Optional[bool] = None,
        entity_ids: Optional[torch.Tensor] = None,
        entity_position_ids: Optional[torch.Tensor] = None,
        entity_embeds: Optional[torch.Tensor] = None,
        **kwargs,
    ):
        return self._forward(
            input_ids=input_ids,
            attention_mask=attention_mask,
            token_type_ids=token_type_ids,
            position_ids=position_ids,
            head_mask=head_mask,
            inputs_embeds=inputs_embeds,
            output_attentions=output_attentions,
            output_hidden_states=output_hidden_states,
            return_dict=return_dict,
            entity_ids=entity_ids,
            entity_position_ids=entity_position_ids,
            entity_embeds=entity_embeds,
            **kwargs,
        )


class KPRModelForXLMRoberta(XLMRobertaPreTrainedModel, KPRMixin):
    config_class = KPRConfigForXLMRoberta

    def __init__(self, config: KPRConfigForXLMRoberta):
        XLMRobertaPreTrainedModel.__init__(self, config)

        self.roberta = XLMRobertaModel(config)
        if self.config.entity_fusion_method != "none":
            self.entity_fusion_layer = EntityFusionLayer(config)

        self.post_init()

    def forward(
        self,
        input_ids: Optional[torch.Tensor] = None,
        attention_mask: Optional[torch.Tensor] = None,
        token_type_ids: Optional[torch.Tensor] = None,
        position_ids: Optional[torch.Tensor] = None,
        head_mask: Optional[torch.Tensor] = None,
        inputs_embeds: Optional[torch.Tensor] = None,
        output_attentions: Optional[bool] = None,
        output_hidden_states: Optional[bool] = None,
        return_dict: Optional[bool] = None,
        entity_ids: Optional[torch.Tensor] = None,
        entity_position_ids: Optional[torch.Tensor] = None,
        entity_embeds: Optional[torch.Tensor] = None,
        **kwargs,
    ):
        return self._forward(
            input_ids=input_ids,
            attention_mask=attention_mask,
            token_type_ids=token_type_ids,
            position_ids=position_ids,
            head_mask=head_mask,
            inputs_embeds=inputs_embeds,
            output_attentions=output_attentions,
            output_hidden_states=output_hidden_states,
            return_dict=return_dict,
            entity_ids=entity_ids,
            entity_position_ids=entity_position_ids,
            entity_embeds=entity_embeds,
            **kwargs,
        )
```
Which will be equivalent (except for being a few lines longer). With this code in place, the following snippet works nicely:
```python
from sentence_transformers import SentenceTransformer
from sentence_transformers.models import Transformer, Pooling, Normalize

transformer = Transformer(
    "knowledgeable-ai/kpr-bge-base-en",
    model_args={"trust_remote_code": True},
    config_args={"trust_remote_code": True},
    tokenizer_args={"trust_remote_code": True},
)
pooling = Pooling(transformer.get_word_embedding_dimension(), pooling_mode="cls")
normalize = Normalize()
model = SentenceTransformer(modules=[transformer, pooling, normalize], trust_remote_code=True)
# Feel free to model.save_pretrained/model.push_to_hub, afterwards you can just use
# model = SentenceTransformer("knowledgeable-ai/kpr-bge-base-en", trust_remote_code=True)

input_texts = ["Who founded Dominican Liberation Party?", "Who owns Mompesson House?"]

embeddings = model.encode(input_texts)

print("Embeddings:", embeddings)
# Embeddings: [[-4.1392539e-03 -4.7670915e-03 -5.3531036e-02 ... -1.9357918e-02
#   -1.3436033e-02  4.1372344e-02]
#  [ 1.5865596e-02  2.4164461e-02 -1.8532677e-02 ...  1.1513681e-02
#    7.7572264e-02 -6.2908999e-05]]
```
Which matches the embeddings from your README.

P.s. if you save the SentenceTransformer model locally, you can see the new files required for the model to load nicely with SentenceTransformer again, e.g.:
* `modules.json` to inform ST which modules the model consists of (e.g. transformer, pooling, normalize)
* An (optional) `config_sentence_transformers.json` with e.g. prompts or the default similarity function (e.g. "cosine")
* An (optional) `sentence_bert_config.json`, the config file informing defaults for the `Transformer` module, with `max_seq_length`/`do_lower_case`.
* A mandatory `1_Pooling/config.json`, the config file for the `Pooling` module, showing CLS pooling (as nicely supported by your model). 

No other files will need to be changed. If you upload these files to your repository, e.g. [`knowledgeable-ai/kpr-bge-base-en`](https://huggingface.co/knowledgeable-ai/kpr-bge-base-en), then the following snippet will also work:
```python
from sentence_transformers import SentenceTransformer

model = SentenceTransformer("knowledgeable-ai/kpr-bge-base-en", trust_remote_code=True)

input_texts = ["Who founded Dominican Liberation Party?", "Who owns Mompesson House?"]

embeddings = model.encode(input_texts)

print("Embeddings:", embeddings)
# Embeddings: [[-4.1392539e-03 -4.7670915e-03 -5.3531036e-02 ... -1.9357918e-02
#   -1.3436033e-02  4.1372344e-02]
#  [ 1.5865596e-02  2.4164461e-02 -1.8532677e-02 ...  1.1513681e-02
#    7.7572264e-02 -6.2908999e-05]]
```

What do you think of this proposal? I.e. the `inspect.signature` which requires some minor changes on your side?

- Tom Aarsen

**Comment by ikuyamada** [comment]
Thanks a lot for your prompt work! 
I agree that we should avoid introducing architecture-specific implementation into this library.

I’ve updated our modeling code to explicitly define the argument names and verified that it works as expected with sentence-transformers. 
So, I’d really appreciate it if you could merge this PR.

Also, thank you for suggesting to host the configuration files on the HF Hub. 
As instructed, we will host it there soon!

**Comment by tomaarsen** [comment]
Perfect, thank you! I think this is a very solid solution. We can ignore the test failures, as of a couple of weeks I've been having issues with rate limits on my GitHub Actions tests. I looked them over, and they look unrelated to the PR.

I'll be merging this now. You can try it out prior to the next Sentence Transformers release already with this:
```
pip install git+https://github.com/UKPLab/sentence-transformers.git
```

- Tom Aarsen

**Comment by ikuyamada** [comment]
Thanks a lot for merging this pull request!
I’ve confirmed that [all KPR models](https://huggingface.co/knowledgeable-ai/models) work with Sentence Transformers😄
Really looking forward to the next release!
